server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java

```diff
@@ -216,6 +225,46 @@ private long removeBlipCandidates(GarbageCollectionEnvironment gce,
     return blipCount;
   }
 
+  @VisibleForTesting
+  /**
+   * Double check no tables were missed during GC
+   */
+  protected void ensureAllTablesChecked(Set<TableId> tableIdsBefore, Set<TableId> tableIdsSeen,
+      Set<TableId> tableIdsAfter) {
```

Comment:
Elsewhere there was a comment about using table states.  If we had tables states from ZK, then something like the following would avoid race conditions.

```suggestion
  protected void ensureAllTablesChecked(Map<TableId,TableState> tableIdsAndStatesBefore, Set<TableId> tableIdsSeen,
      Map<TableId,TableState> tableIdsAndStatesAfter) {
      
      //TODO maybe copy maps before deleting from them
      //only consider tables states ONLINE and OFFLINE because entries must exist in the metadata table when a table is in these states.  Look at how table states are set relative to metadata updates in table creation and deletion for more details.
      tableIdsAndStatesBefore.values().removeIf(tabState - > tabState != ONLINE && tabState != OFFLINE);
      tableIdsAndStatesAfter.values().removeIf(tabState - > tabState != ONLINE && tabState != OFFLINE);
      
      var tableIdsBefore = tableIdsAndStatesBefore.keySet();
      var tableIdsAfter = tableIdsAndStatesAfter.keySet();
```

Reply:
After looking at the javadoc that @ctubbsii  said was confusing and trying to improve it, i think the filtering on table state should be done in `getCandidateTableIDs()`.  So `getCandidateTableIDs()` only returns table ids where the state is ONLINE or OFFLINE in its impl.  Then ensureAllTablesChecked would continue to take sets and stay the same, not taking the maps as I suggested earlier.

Reply:
I think I got this covered in 1f8d37305789fa1540158e32b54da86a7af3ae87. Take a look if you can.